spring-kafka/src/main/java/org/springframework/kafka/core/DefaultTransactionSuffixManager.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2016-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core;
+
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link TransactionSuffixManager} implementation for managing transactional producer suffixes.
+ * If the {@link #maxCache} is greater than 0, the suffixes will be cached and reused.
+ * Otherwise, the suffixes will be generated on the fly.
+ *
+ * <p>
+ * Setting the {@link #setMaxCache(int)} enables caching to restrict the number of `transactional.id`.
+ * The cache is per `transactional.id` prefix. The cache size is limited by the {@link #maxCache}.
+ * Default is 0, which means no caching and restriction, so the `transactional.id` will be generated on the fly.
+ *
+ * @author Ilya Starchenko
+ *
+ * @since 3.1.1
+ */
+public class DefaultTransactionSuffixManager implements TransactionSuffixManager {
+
+	private final AtomicInteger transactionIdSuffix = new AtomicInteger();
+
+	private final Map<String, BlockingQueue<String>> suffixCache = new ConcurrentHashMap<>();
+
+	private int maxCache;
+
+	/**
+	 * Retrieve the suffix for the transactional producer from the cache or generate a new one 
+	 * if caching is disabled.
+	 * @param txIdPrefix the transaction id prefix.
+	 * @return the suffix.
+	 * @throws NoProducerAvailableException if caching is enabled and no suffixes are available.
+	 */
+	@Override
+	public String retrieveSuffix(String txIdPrefix) {
+		BlockingQueue<String> cache = getSuffixCache(txIdPrefix);
+		if (cache == null) {
+			return String.valueOf(this.transactionIdSuffix.getAndIncrement());
+		}
+
+		String suffix = cache.poll();
+		if (suffix == null) {
+			throw new NoProducerAvailableException("No available transaction producer suffix for " + txIdPrefix);
+		}
+		return suffix;
+	}
+
+	@Override
+	public void returnSuffix(@Nullable String txIdPrefix, @Nullable String suffix) {
```

Comment:
We need epoch when return suffix, prev epoch producer  maybe close after reset at next epoch.

Reply:
I don't quite understand how `epoch` and `txId` are related in this case. If `epoch` is bumped, then the previous producer will be closed, and the suffix will be returned back.

Reply:
such as https://github.com/spring-projects/spring-kafka/blob/2306ac5cbd47702f9a492541186944682edada9c/spring-kafka/src/test/java/org/springframework/kafka/core/DefaultKafkaProducerFactoryTests.java#L257-L288

Reply:
I still don't get it. When `ProducerFactory#reset()` is called, **every** producer is closed, so calling close on the producer itself via `Producer#close()` will do nothing. Am I mistaken?

Reply:
I got what you meant, so I we do not need to return `suffix` back if `epoch` if bumped, because all `suffixes` already returned.

Reply:
Have a look at https://github.com/spring-projects/spring-kafka/blob/2306ac5cbd47702f9a492541186944682edada9c/spring-kafka/src/main/java/org/springframework/kafka/core/DefaultKafkaProducerFactory.java#L762-L788

At line L763-L765, tx producer is diff from nomal producer,
tx producer is not managed by `DefaultKafkaProducerFactory`, so if producer is in transaction when call `DefaultKafkaProducerFactory#reset()`, producer will not close.